### PR TITLE
[feat]: initialized the project

### LINKED ISSUE #1: Welcoming initial PoC PR

### PR DESCRIPTION
**Pull request by Palanikannan1437**
## Description

- Closes #1 

This pull request implements a new feature to fetch Hugging Face model URLs, build a data structure, and download the model weights using the Fastify framework.

### Changes

1. **Fastify Implementation**
   - Set up a Fastify project.
   - Created a Fastify app that fetches a specified webpage and extracts Hugging Face URLs using the `linkify-it` package.
   - Built a data structure based on the extracted URLs.
   - Implemented the `downloadFromHub` function to download the model weights using the `@huggingface/hub` library and saved them to a `weights` directory with each of the extracted models in their subdirectory recursively using `@huggingface/hub` --> This doesn't come straighforward i.e. first using the repoName, on has to fetch all the files and then recursively traverse to write it into the filesystem.
   - Handled file and directory operations using the `fs` module.
   - Improved error handling and used `Promise.allSettled` to handle multiple downloads concurrently.
      - The `downloadPromises` array is created by mapping over the models and returning a promise for each download.
      - Each promise resolves with a success message if the download is successful or rejects with an error message if the download fails.
      - `Promise.allSettled` is used to wait for all download promises to settle (either fulfilled or rejected).
      - The results are filtered to count the number of failed downloads.
      - If there are any failed downloads, an error message is sent as the response, otherwise, a success message is sent.
      - If the request is terminated in between, then there are checks to not redownload that specific file if already present in the filesystem.

### Dependencies

- Fastify
- linkify-it
- @huggingface/hub
- a minimal logger (chalk)

### Testing

To test the Fastify app, run the following command:

```bash
npm run start
```

## Summary

The Fastify app would fetch the specified webpage, extract Hugging Face URLs, build a data structure, and download the model weights using the `downloadFromHub` function. The downloaded weights would be saved in the target directory. The app should handle errors and use `Promise.allSettled` to manage multiple downloads concurrently.

## Demo
Sep-01-2023 18-46-21

### CODE REVIEW

**Comment by Palanikannan1437** [comment]
@Sing-Li would love your review on this, thanks 🚀

File: src/index.mjs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+const server = fastify()
+
+server.get('/', async (request, reply) => {
+  try {
+    const huggingFaceURLs = await fetchPageAndExtractURLs();
+    const dataStructure = buildDataStructure(huggingFaceURLs);
+    log('info', 'Starting model downloads');
+
+    const downloadPromises = dataStructure.models.map(model => {
```
Comment (Sing-Li) [inline src/index.mjs:21]: Can we put some concurrency control here to get better performance?     Like limited max_concurrent_downloads to    a number, say 3 or 10     ....  perhaps in a dotev?

**Review by Sing-Li — CHANGES_REQUESTED** [review]
@Palanikannan1437  please add concurrency control

File: src/index.mjs (hunk 1 above)
Comment (Palanikannan1437) [reply src/index.mjs:21]: Done!! 🚀

**Review by Sing-Li — APPROVED** [review]
awesome work @Palanikannan1437 !   Good enough for initial proof of concept commit. congrats.

only comment is that it is a bit overweight, several libs with tons of dependencies for the skeletal code at this point.      good enough though, we can slim it down over time if we have to.

**Pull request merged**